File: 이진영/백준/10835.카드게임.js
// 검색 풀이

var input = require('fs').readFileSync("dev/stdin").toString().trim().split("\n");
var N = +input[0].trim();
var left = input[1].trim().split(" ").map(Number);
var right = input[2].trim().split(" ").map(Number);

var dp = Array.from({ length: N }, () => Array(N).fill(-1));

var f = function (l, r) {
    if (l === N || r === N) return 0;
    if (dp[l][r] !== -1) return dp[l][r];

    dp[l][r] = Math.max(f(l+1, r+1), f(l+1, r));

    if (left[l] > right[r]) {
        dp[l][r] = Math.max(dp[l][r], f(l, r+1) + right[r]);
    }
    return dp[l][r];
}

console.log(f(0, 0));

File: 김채린/추가문제/re백준/10835_카드게임.java
import java.util.*;
import java.io.*;
class Main{
    static int n;
    static ArrayList<String> result;
    static int[] a;
    static int [] b;
    static int[][] dp;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        dp = new int[n][n];
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], -1);
        }
        a = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        b = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        sol(0,0,0);
        System.out.print(dp[0][0]);
    }

    public static int sol(int left, int right, int total) {
        if(left==n||right==n){
            return 0;
        }
        if(dp[left][right]!=-1){
            return dp[left][right];
        }
        int result1 =sol(left+1, right,total);
        int result2 =sol(left+1,right+1,total);
        int result3 = total;
        if (a[left] > b[right]) {
            result3 = sol(left, right + 1, total);
            result3 += b[right];
        }
        dp[left][right] = Math.max(result1, Math.max(result2, result3));
        return dp[left][right];
    }
}
